parsers/AU_TAS_KI.py

```diff
@@ -0,0 +1,81 @@
+        uri = "wss://data.ajenti.com.au/live/signalr/reconnect?transport=webSockets&messageId=d-C8F39B96-km%2C152%7Ckr%2C0%7Cks%2C2&clientProtocol=1.5&connectionToken=Yu0BMoSk5rEBlaiSH%2BaLhoYTA3oiXDBcS%2FvqcVQr7q%2FnKYKbTppUSNS1Mz%2Bx1wwgH4VHAgSv5SalVBzsYrWiG92AM7r5qBJogpCJYS%2BqEWECTRQS%2BUl62HQ5UjFC6zCh&connectionData=%5B%7B%22name%22%3A%22taghub%22%7D%5D&tid=3"
+        async with websockets.connect(uri) as websocket:
+            payload_str = await websocket.recv()
+            payload = json.loads(payload_str)
+            return payload
+
+    # Sometimes the webservice returns an empty dic (in ~20% of the calls)... Let's retry two times in those cases
+    res = {}
+    tries = 0
+    while res == {} and tries < 2:
+        res = asyncio.get_event_loop().run_until_complete(fetch())
+        tries+=1
+    return res
+
+def parse_payload(logger, payload):
+    for message in payload['M']:
+        for serie in message['A']:
+            logger.debug(f"serie : {json.dumps(serie)}")
+            for technology in serie['technologies']:
+                assert technology['unit'] == 'kW'
+                # The upstream API gives us kW, we need MW
+                technologies_parsed[technology['id']] = int(technology['value'])/1000
+            logger.debug(f"production : {json.dumps(technologies_parsed)}")
+
+            biodiesel_percent = serie['biodiesel']['percent']
+
+
+    return technologies_parsed, biodiesel_percent
+
+def fetch_production(zone_key='AUS-TAS-KI', session=None, target_datetime=None, logger: logging.Logger = logging.getLogger(__name__)):
+
+    if target_datetime is not None:
+        raise NotImplementedError('The datasource currently implemented is only real time')
+  
+    if session is not None:
+        # TODO : is that the best to do in our case ? (websockets)
+        raise NotImplementedError
+    
+    payload = fetch_api()
+    technologies_parsed, biodiesel_percent = parse_payload(logger, payload)
+
+    """ 
+    TODO
+        * What do I return for production mode that are not even on the island ? 
+            The main README says "The production values should never be negative. Use None, or omit the key if a specific production mode is not known." but doesn't answer this question
+    """
+    return {
+      'zoneKey': zone_key,
+      'datetime': arrow.now(tz='Australia/Currie').datetime,
+      'production': {
+          'biomass': technologies_parsed['diesel']*biodiesel_percent/100,
+          'coal': 0,
+          'gas': 0,
+          'hydro': 0,
+          'nuclear': 0,
+          'oil': technologies_parsed['diesel']*(100-biodiesel_percent)/100,
+          'solar': technologies_parsed['solar'],
+          'wind': technologies_parsed['wind'],
+          'geothermal': 0,
+          'unknown': 0
+      },
+      'storage': {
+          'battery': technologies_parsed['battery']*-1,
+          'flywheel': technologies_parsed['flywheel']*-1,
```

Comment:
I've [searched](https://github.com/tmrowco/electricitymap-contrib/search?q=flywheel&type=Code) the project for any mentions of _flywheel_, and besides the original King island issue nothing comes up. I think that either this needs to be grouped with _battery_, or a new storage category might have to be added.

Probably worth asking someone from Tomorrow what the preferred course of action is.

Reply:
Do you know who would be the best to @ to get an opinion about that ?

Reply:
1. I think adding the flywheel to "battery" is the best option here.
It is used for "very short time storage and production" / for frequency balancing on the island in combo with the shunt resistor (which is an additional load to suppress overfrequency at times). The frequency of the changes in energy direction of the flywheel is of course very high. I'll add a more general comment below regarding that problem
2. "Battery" is "OFF" at the moment but consuming "3kW". In this case, -3kW don't go into the battery itself but probably are consumed for the auxiliary devices and controllers. @robertahunt is this something you'd set to 0 kW when "off" or keep the consumption to account for stand-by emissions?

Reply:
@alixunderplatz in relation to your (1), please note that I didn't do anything with the "resistor" part of [the monitoring](https://data.ajenti.com.au/KIREIP/index.html). I see it as a way to "burn" the excess electricity generated for grid regulation. Maybe the flywheel should be considered like the resistor and just not counted at all ?

Reply:
Agreed with @alixunderplatz that flywheel could be included under 'battery' to simplify things. And in general for the battery, I think it's numbers should be split across storage and production - when it is charging, this should be included as a negative value under 'storage', and when it is discharging this should be a positive value under production - This is what we have done with hydro storage systems for example in EIA.py and JP_ISEP.py

Reply:
> @robertahunt How often will the parser collect the data? The source data page refreshes every few seconds, wouldn't it make sense to collect 10-15 datapoints to put them into an average for all the types and return the averaged values? This is the most-real-time source that we have, with the most variable production as well. As the system size is very small (up to 3 MW) and as it has a high share of variable sources in it, the frequency of the mini-grid will not be as stable, causing each fetch_production to be a snapshot of the last few seconds and not a timespan of a minute or more. Especially when it comes to storage.

I am not sure I understand perfectly what is happening here, so correct me if my response doesn't seem to jive. The parsers run approximately every 15 minutes, but we are happy to get the highest granularity of data possible, so if data is available at a 1 minute frequency, we are happy to keep all those datapoints then our processor will take the average.